### ISSUE TITLE
is this module thread safe?

### ISSUE THREAD

**Issue by obilodeau** [briandfoy/crypt-rijndael#2]
I've seen a segfault in the perl vm on a multi-threaded daemon that uses Crypt::Rijndael (through Net::SNMP actually).

At this point I don't know for sure by what it's caused by I thought I would save some time by asking you if you knew if the module is thread safe or not since it's not explicitly documented.

Here's the stacktrace:

```
*** glibc detected *** /usr/bin/perl: double free or corruption (fasttop): 0x9e70bc38 ***
... one line omitted ...
======= Backtrace: =========
/lib/libc.so.6[0xa666c5]
/lib/libc.so.6(cfree+0x59)[0xa66b09]
/usr/lib/perl5/5.8.8/i386-linux-thread-multi/CORE/libperl.so(Perl_newCONSTSUB+0x154)[0x150354]
/usr/lib/perl5/vendor_perl/5.8.8/i386-linux-thread-multi/auto/Crypt/Rijndael/Rijndael.so(boot_Crypt_
_Rijndael+0x491)[0x9
ce201]
/usr/lib/perl5/5.8.8/i386-linux-thread-multi/CORE/libperl.so(Perl_pp_entersub+0x40d)[0x18952d]
/usr/lib/perl5/5.8.8/i386-linux-thread-multi/CORE/libperl.so(Perl_runops_standard+0x1f)[0x18299f]
/usr/lib/perl5/5.8.8/i386-linux-thread-multi/CORE/libperl.so[0x12303e]
/usr/lib/perl5/5.8.8/i386-linux-thread-multi/CORE/libperl.so(Perl_call_sv+0x5e6)[0x127906]
/usr/lib/perl5/5.8.8/i386-linux-thread-multi/auto/threads/threads.so(Perl_ithread_run+0x1ae)[0x968a5
e]
/lib/libpthread.so.0[0xba4832]
/lib/libc.so.6(clone+0x5e)[0xacf45e]
```

Thanks,
Big fan of your work btw.

update: here's a link to the [downstream issue][1]
[1]: http://packetfence.org/bugs/view.php?id=1378

**Comment by Leont** [comment]
Looks like a subtle issue, that isn't so much about thread-safety but apparently can be triggered by it (how that suddenly happens to be triggered isn't entirely obvious to me). Should be easy to fix, but a test-case would be helpful @obilodeau.

**Comment by Leont** [comment]
Actually, it seems this module is not thread safe, the cipher objects are not cloned on thread-clone, so they can easily get double-freed and double-used!

**Comment by obilodeau** [comment]
I'm sorry I have no test case. I don't work at Inverse anymore. The bug was trigerred when doing multi-threaded SNMPv3 network I/O with switches configured to use AES as their cipher. That's all I recall.

**Comment by Leont** [comment]
This issue might be solved by 1.13, or may not. Lacking a test-case I'm not sure.
